src/plugins/workspace/public/components/workspace_menu/workspace_menu.tsx

```diff
@@ -3,62 +3,73 @@
   EuiFlexGroup,
   EuiFlexItem,
-  EuiIcon,
-  EuiListGroup,
-  EuiListGroupItem,
+  EuiPanel,
   EuiPopover,
   EuiText,
 } from '@elastic/eui';
-import type { EuiContextMenuPanelItemDescriptor } from '@elastic/eui';
-
+import { FormattedMessage } from '@osd/i18n/react';
+import { BehaviorSubject } from 'rxjs';
 import {
   WORKSPACE_CREATE_APP_ID,
   WORKSPACE_LIST_APP_ID,
+  MAX_WORKSPACE_PICKER_NUM,
   WORKSPACE_DETAIL_APP_ID,
 } from '../../../common/constants';
-import { cleanWorkspaceId, formatUrlWithWorkspaceId } from '../../../../../core/public/utils';
+import { formatUrlWithWorkspaceId } from '../../../../../core/public/utils';
 import { CoreStart, WorkspaceObject } from '../../../../../core/public';
+import { getUseCaseFromFeatureConfig } from '../../utils';
+import { recentWorkspaceManager } from '../../recent_workspace_manager';
+import { WorkspaceUseCase } from '../../types';
 
+const defaultHeaderName = i18n.translate(
+  'core.ui.primaryNav.workspacePickerMenu.defaultHeaderName',
+  {
+    defaultMessage: 'Workspaces',
+  }
+);
 interface Props {
   coreStart: CoreStart;
+  registeredUseCases$: BehaviorSubject<WorkspaceUseCase[]>;
 }
 
-/**
- * Return maximum five workspaces, the current selected workspace
- * will be on the top of the list.
- */
-function getFilteredWorkspaceList(
-  workspaceList: WorkspaceObject[],
-  currentWorkspace: WorkspaceObject | null
-): WorkspaceObject[] {
-  return [
-    ...(currentWorkspace ? [currentWorkspace] : []),
-    ...workspaceList.filter((workspace) => workspace.id !== currentWorkspace?.id),
-  ].slice(0, 5);
-}
-
-export const WorkspaceMenu = ({ coreStart }: Props) => {
+export const WorkspaceMenu = ({ coreStart, registeredUseCases$ }: Props) => {
   const [isPopoverOpen, setPopover] = useState(false);
   const currentWorkspace = useObservable(coreStart.workspaces.currentWorkspace$, null);
   const workspaceList = useObservable(coreStart.workspaces.workspaceList$, []);
+  const isDashboardAdmin = !!coreStart.application.capabilities.dashboards;
+  const availableUseCases = useObservable(registeredUseCases$, []);
+
+  const filteredWorkspaceList = useMemo(() => {
+    return workspaceList.slice(0, MAX_WORKSPACE_PICKER_NUM);
+  }, [workspaceList]);
+
+  const filteredRecentWorkspaces = useMemo(() => {
+    return recentWorkspaceManager
+      .getRecentWorkspaces()
+      .map((workspace) => workspaceList.find((ws) => ws.id === workspace.id))
+      .filter((workspace): workspace is WorkspaceObject => workspace !== undefined)
+      .slice(0, MAX_WORKSPACE_PICKER_NUM);
+  }, [workspaceList]);
 
-  const defaultHeaderName = i18n.translate(
-    'core.ui.primaryNav.workspacePickerMenu.defaultHeaderName',
-    {
-      defaultMessage: 'Select a workspace',
-    }
-  );
-  const filteredWorkspaceList = getFilteredWorkspaceList(workspaceList, currentWorkspace);
   const currentWorkspaceName = currentWorkspace?.name ?? defaultHeaderName;
 
+  const getUseCase = (workspace: WorkspaceObject) => {
```

Comment:
Not sure if Ling's PR will affect the code here.

Reply:
I have already talked to Lin, it will not affect my code. It's just each using our own. 😹

Reply:
updated